lib/internal/modules/esm/fetch_module.js

```diff
@@ -99,118 +96,75 @@ function fetchWithRedirects(parsed) {
     return existing;
   }
   const handler = parsed.protocol === 'http:' ? HTTPGet : HTTPSGet;
-  const result = new Promise((fulfill, reject) => {
+  const result = (async () => {
     const req = handler(parsed, {
-      headers: {
-        Accept: '*/*'
-      }
-    })
-    .on('error', reject)
-    .on('response', (res) => {
-      function dispose() {
-        req.destroy();
-        res.destroy();
-      }
+      headers: { Accept: '*/*' },
+    });
+    const { 0: res } = await once(req, 'response');
+    try {
       if (res.statusCode >= 300 && res.statusCode <= 303) {
-        if (res.headers.location) {
-          dispose();
-          try {
-            const location = new URL(res.headers.location, parsed);
-            if (location.protocol !== 'http:' &&
-              location.protocol !== 'https:') {
-              reject(new ERR_NETWORK_IMPORT_DISALLOWED(
-                res.headers.location,
-                parsed.href,
-                'cannot redirect to non-network location'));
-              return;
-            }
-            return PromisePrototypeThen(
-              PromiseResolve(fetchWithRedirects(location)),
-              (entry) => {
-                cacheForGET.set(parsed.href, entry);
-                fulfill(entry);
-              });
-          } catch (e) {
-            dispose();
-            reject(e);
-          }
+        const location = new URL(res.headers.location, parsed);
+        if (location.protocol !== 'http:' && location.protocol !== 'https:') {
+          throw new ERR_NETWORK_IMPORT_DISALLOWED(
+            res.headers.location,
+            parsed.href,
+            'cannot redirect to non-network location'
+          );
         }
+        const entry = await fetchWithRedirects(location);
+        cacheForGET.set(parsed.href, entry);
+        return entry;
+      }
+      if (res.statusCode === 404) {
+        const err = new ERR_MODULE_NOT_FOUND(parsed.hred);
+        err.message = `Cannot find module '${parsed.href}', HTTP 404`;
+        throw err;
       }
       if (res.statusCode > 303 || res.statusCode < 200) {
-        dispose();
-        reject(
-          new ERR_NETWORK_IMPORT_BAD_RESPONSE(
-            parsed.href,
-            'HTTP response returned status code of ' + res.statusCode));
-        return;
+        throw new ERR_NETWORK_IMPORT_DISALLOWED(
```

Comment:
please dispose/close out the connection here and other places so a dangling response doesn't just keep itself open.

Reply:
dispose() was doing this but idk why it was removed.

Reply:
@bmeck basically because we have better/newer/nicer resource management facilities:
 - on requests - closing the request closes the response 
 - on the body stream - using `compose` ensures all the underlying streams are closed.

this is why it's enough to close the request/bodyStream in `finally`clauses to get the same resource handling :)